### PR TITLE
adding bin packing problem

### PR DESCRIPTION
**Pull request by alejomonbar**
[Unitary Fund](http://unitary.fund)

Adding the bin packing problem using the docplex template. Additionally, solving some issues with the class FromDocplexToIsingModel.

## Description
 I would like to unify the different optimization problems in ./openqaoa/problems defining them as docplex models.
I've presented the bin packing problem using this new methodology.

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [ ] My changes generate no new warnings
- [ ] I have added/updated tests to make sure bugfix/feature works.
- [ ] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Name the new unit-tests that you have added along with this change.

### CODE REVIEW

File: openqaoa/problems/converters.py [hunk 1]
```diff
@@ -317,14 +323,14 @@ def linear_constraints(self, multipliers=None) -> None:
             ]:  # Inequality constraint added as a penalty with additional slack variables.
                 constraint.name = f"C{cn}"
                 if self.unbalanced:
-                    penalty = multipliers[cn] * self.inequality_to_unbalanced_penalty(
+                    penalty = self.inequality_to_unbalanced_penalty(
```
Comment (shahidee44) [inline openqaoa/problems/converters.py:326]: Why is multipliers removed from this formula.

File: openqaoa/problems/binpacking.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        else:
+            solution = self.solution.copy()
+        for var in self.cplex_model.iter_binary_vars():
+            if string:
+                solution += str(int(np.round(docplex_sol.get_value(var), 1)))
+            else:
+                solution[var.name] = int(np.round(docplex_sol.get_value(var), 1))
+        if not docplex_sol.is_valid_solution():
+            raise ValueError("The Cplex solver does not find a solution.")
+        return solution
```
Comment (shahidee44) [inline openqaoa/problems/binpacking.py:171]: This function looks generalizable for any cplex_model object. As mentioned, maybe its better for this method to be a function in the utilities class where we do something like `solve_with_cplex`

File: openqaoa/problems/converters.py (hunk 1 above)
Comment (alejomonbar) [reply openqaoa/problems/converters.py:326]: I decided to follow the structure of the paper (https://arxiv.org/abs/2211.13914). In this case, I don't need to multiply this penalization term (which comes from the equality constraints) and the unbalanced penalization term. I split this pull request into two, the part where I covered the problems with `FromDocplex2IsingModel` are in #198.

File: openqaoa/problems/binpacking.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        self.method = method
+        self.simplifications = simplifications
+        self.include_ineqs = include_ineqs
+        self.eq_constraints = {}
+        self.ineq_constraints = {}
+        if n_bins is None:
+            self.n_bins = self.n_items
+        else:
+            self.n_bins = n_bins
+        if len(weights) > 0:
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:69]: I removed the option to check if there are elements in the weight list (#201 ). Now, having an empty list is an option.

File: openqaoa/problems/binpacking.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+                        self.ineq_constraints[f"ineq_{j}"] = [[self.vars_pos[f"x_{i}_{j}"] for i in list_items], [self.weight_capacity - self.weights[0]]]
+                    else:
+                        self.ineq_constraints[f"ineq_{j}"] = [[self.vars_pos[f"x_{i}_{j}"] for i in list_items], [self.weight_capacity]]
+                        
+                else:
+                    self.ineq_constraints[f"ineq_{j}"] = [[self.vars_pos[f"x_{i}_{j}"] for i in list_items], [self.vars_pos[f"y_{j}"]]]
+
+        return mdl
+
+    def get_qubo_problem(self):
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:142]: Done! #201

File: openqaoa/problems/binpacking.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            raise ValueError("The Cplex solver does not find a solution.")
+        return solution
+    
+    def plot_solution(self, solution:Union[dict, str], ax=None):
+        if isinstance(solution, str):
+            sol = self.solution.copy()
+            for n, var in enumerate(self.cplex_model.iter_binary_vars()):
+                sol[var.name] = int(solution[n])
+            solution = sol
+        colors = plt.cm.get_cmap("jet", len(self.weights))
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:184]: Let me know if I should modify it with a new palette of colors.

File: openqaoa/problems/binpacking.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        An instance of the Bin PAcking problem.
+    """
+
+    __name__ = "binpacking"
+    
+    def __init__(self, weights:list=[], weight_capacity:int=0, penalty:Union[float, list]=None,
+                 n_bins:int = None, simplifications=True, method="slack", include_ineqs:bool=True):
+        #include_ineqs: True if including the inequalities
+
+        self.weights = weights
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:57]: Done!

File: openqaoa/problems/binpacking.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        n_betas = len(betas)
+        n_gammas = len(gammas)
+        landscape =[ ]
+        for params in itertools.product(betas, gammas):
+            landscape.append(self.cost_energy(params, qaoa_mdl))
+        landscape = np.array(landscape)
+        landscape = landscape.reshape((n_betas, n_gammas))
+        return landscape
+
+    def cost_energy(self, params, qaoa):
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:300]: I do not include this functionality anymore. Maybe, in the future, we can allocate it to another part.

File: openqaoa/problems/binpacking.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+       varational_params = qaoa.variate_params
+       varational_params.update_from_raw(params)
+       return qaoa.backend.expectation(varational_params) 
+
+    def optimal_prob(self, params, qaoa, optimal):
+        varational_params = qaoa.variate_params
+        varational_params.update_from_raw(params)
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:307]: same of https://github.com/entropicalabs/openqaoa/pull/181#discussion_r1138375623

File: openqaoa/problems/binpacking.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    
+    def normalizing(self, problem, normalized=-1):
+        abs_weights = np.unique(np.abs(problem.weights))
+        arg_sort = np.argsort(abs_weights)
+        max_weight = abs_weights[arg_sort[normalized]]
+        weights = [weight / max_weight for weight in problem.weights]
+        terms = problem.terms
+        terms.append([])
+        weights.append(problem.constant/ max_weight)
+        return QUBO(self.n_vars, terms, weights)
```
[the remark this answers is not in the document]
Comment (alejomonbar) [reply openqaoa/problems/binpacking.py:319]: I don't include this function anymore.

**Comment by vishal-ph** [comment]
@alejomonbar, there seems to be a merge conflict in a file

**Comment by alejomonbar** [comment]
> @alejomonbar, there seems to be a merge conflict in a file

@vishal-ph This is no anymore a pull request. I divided it into two different pull requests #201 and #198

**Pull request closed without merging**